### DISCUSSION TITLE
For people who want to contribute to the C extension, how do we test changes locally?

### DISCUSSION THREAD

**Discussion by hirehamir** [microsoft/vscode-cpptools#8745] [Q&A] [answered]
For people who want to contribute to the C extension, how do we start the extension on our local devices in development mode?

For example, usually we can press `F5` to test some Visual Studio extension locally. In which directory would we press `F5` to test our local changes for the C extension?

**Comment by sean-mcmanus** [comment, marked as answer]
https://github.com/microsoft/vscode-cpptools/tree/main/Extension from https://github.com/microsoft/vscode-cpptools/blob/main/Documentation/Building%20the%20Extension.md from https://github.com/microsoft/vscode-cpptools/blob/main/CONTRIBUTING.md .

But FYI, the main branch assumes a binary from 1.8.1.

**Reply by hirehamir** [reply]
Thank you for your response!

How do I get a binary from 1.8.1?

I get this error after pressing `F5` in the `Extension` subdirectory.

`Unable to start the C/C++ language server. IntelliSense features will be disabled. Error: Missing binary at /home/hamir/repositories/vscode-cpptools/Extension/bin/cpptools`

![image](https://user-images.githubusercontent.com/43425812/152034248-0f4d1a32-1ff6-456b-8ee5-df4341030b22.png)

I don't have a `cpptools` subdirectory in `vscode-cpptools/Extension/bin/`.
```
hamir:~$ cd /home/hamir/repositories/vscode-cpptools/Extension/bin/
hamir:~/repositories/vscode-cpptools/Extension/bin$ ls
common.json             macos.clang.arm64.json    windows.clang.x64.json
linux.clang.arm.json    macos.clang.x64.json      windows.clang.x86.json
linux.clang.arm64.json  macos.clang.x86.json      windows.gcc.arm.json
linux.clang.x64.json    macos.gcc.arm.json        windows.gcc.arm64.json
linux.clang.x86.json    macos.gcc.arm64.json      windows.gcc.x64.json
linux.gcc.arm.json      macos.gcc.x64.json        windows.gcc.x86.json
linux.gcc.arm64.json    macos.gcc.x86.json        windows.msvc.arm.json
linux.gcc.x64.json      messages                  windows.msvc.arm64.json
linux.gcc.x86.json      windows.clang.arm.json    windows.msvc.x64.json
macos.clang.arm.json    windows.clang.arm64.json  windows.msvc.x86.json
```

I ran `git clone -b release https://github.com/Microsoft/vscode-cpptools` and `yarn install` without any issues as mentioned in, https://github.com/microsoft/vscode-cpptools/blob/main/Documentation/Building%20the%20Extension.md. 

The error persists after deleting `yarn.lock` and running `yarn install` again.

```
hamir:~/repositories/vscode-cpptools/Extension$ yarn install
yarn install v1.22.17
info No lockfile found.
[1/5] Validating package.json...
warning cpptools@1.7.1-main: The engine "vscode" appears to be invalid.
[2/5] Resolving packages...
warning vscode-extension-telemetry@0.4.5: This package has been renamed to @vscode/extension-telemetry, please update to the new name
...
warning vscode-debugprotocol@1.51.0: This package has been renamed to @vscode/debugprotocol, please update to the new name
[3/5] Fetching packages...
warning vscode-cpptools@5.0.0: The engine "vscode" appears to be invalid.
warning vscode-extension-telemetry@0.4.5: The engine "vscode" appears to be invalid.
warning vscode-languageclient@5.2.1: The engine "vscode" appears to be invalid.
warning vscode-tas-client@0.1.27: The engine "vscode" appears to be invalid.
[4/5] Linking dependencies...
[5/5] Building fresh packages...
success Saved lockfile.
Done in 21.99s.
hamir:~/repositories/vscode-cpptools/Extension$ 
```

**Reply by hirehamir** [reply]
Also, perhaps unrelated, I get an error for `fs.readFile` when I press `F5`.

![image](https://user-images.githubusercontent.com/43425812/152036950-b4ee460e-bbf3-4d46-ad1a-ad2849144834.png)

```
No overload matches this call.
  Overload 1 of 4, '(path: number | PathLike, options: { encoding?: null | undefined; flag?: string | undefined; } | null | undefined, callback: (err: ErrnoException | null, data: Buffer) => void): void', gave the following error.
    Argument of type 'string' is not assignable to parameter of type '{ encoding?: null | undefined; flag?: string | undefined; } | null | undefined'.
  Overload 2 of 4, '(path: number | PathLike, options: BufferEncoding | { encoding: BufferEncoding; flag?: string | undefined; }, callback: (err: ErrnoException | null, data: string) => void): void', gave the following error.
    Argument of type 'string' is not assignable to parameter of type 'BufferEncoding | { encoding: BufferEncoding; flag?: string | undefined; }'.
  Overload 3 of 4, '(path: number | PathLike, options: BufferEncoding | (BaseEncodingOptions & { flag?: string | undefined; }) | null | undefined, callback: (err: ErrnoException | null, data: string | Buffer) => void): void', gave the following error.
    Argument of type 'string' is not assignable to parameter of type 'BufferEncoding | (BaseEncodingOptions & { flag?: string | undefined; }) | null | undefined'.
```

I'm able to circumvent it by changing `encoding` to `"utf8"`, but then I run into the other `cpptools` error I mentioned above.

```
hamir:~/repositories/vscode-cpptools/Extension$ tsc -v
Version 4.5.2
hamir:~/repositories/vscode-cpptools/Extension$ node -v
v16.13.0
```

**Reply by sean-mcmanus** [reply]
Prior to 1.8.0, our extension would automatically download offline binaries, but we've stopped publishing those.

You can get the 1.8.2 binaries from installing the extension and then copying the binaries. Are you able to get that working?

**Reply by hirehamir** [reply]
I believe that did the trick! Thanks for taking the time to clarify this. I expanded `Documentation/Building the Extension.md` in #8788 to hopefully help any future developers who might bump into this issue.

**Reply by JominL** [reply]
Is the bin/cpptools close source?

**Reply by JoshuaCF** [reply]
> Prior to 1.8.0, our extension would automatically download offline binaries, but we've stopped publishing those.
> 
> You can get the 1.8.2 binaries from installing the extension and then copying the binaries. Are you able to get that working?

I want to use the language server of the extension with Neovim in WSL. Is there not a URL somewhere I can just download directly from?

**Reply by bobbrow** [reply]
> Is the bin/cpptools close source?

Yes.

> I want to use the language server of the extension with Neovim in WSL. Is there not a URL somewhere I can just download directly from?

Our language server binaries are not licensed to be used outside of the Visual Studio family of products. You can download them if you are contributing to the open source TypeScript portion of the code, but they aren't licensed to be used in a vim extension.
